File: src/movil/lib/auth/jwt.dart
import 'dart:convert';

import '../api/modelos.dart';

/// Saca el usuario de los claims del access token.
///
/// Provisional, igual que en la web: cuando el backend agregue GET /api/auth/me
/// o mande el usuario en la respuesta del login, esto se reemplaza.
UsuarioSesion? leerUsuarioDelToken(String token) {
  final partes = token.split('.');
  if (partes.length < 2) {
    return null;
  }

  try {
    final cuerpo = utf8.decode(base64Url.decode(base64Url.normalize(partes[1])));
    final claims = jsonDecode(cuerpo) as Map<String, dynamic>;
    final id = claims['sub'] as String?;
    if (id == null) {
      return null;
    }

    final email = claims['email'] as String? ?? '';
    return UsuarioSesion(
      id: id,
      email: email,
      nombre: claims['name'] as String? ?? (email.isEmpty ? 'Usuario' : email),
    );
  } catch (_) {
    return null;
  }
}

File: src/web/src/auth/jwt.ts
export type UsuarioSesion = {
  id: string
  email: string
  nombre: string
}

type ClaimsToken = {
  sub?: string
  email?: string
  name?: string
}

/**
 * Saca los datos del usuario del access token.
 *
 * Provisional: el backend no expone `GET /api/auth/me` y el token solo trae los
 * claims sub, email, name y jti. En cuanto Paolo agregue el usuario con sus
 * roles a la respuesta del login, esto se reemplaza por ese dato.
 */
export function leerUsuarioDelToken(token: string): UsuarioSesion | null {
  const cuerpo = token.split('.')[1]
  if (!cuerpo) {
    return null
  }

  try {
    const base64 = cuerpo.replace(/-/g, '+').replace(/_/g, '/')
    const texto = decodeURIComponent(
      atob(base64)
        .split('')
        .map((caracter) => `%${caracter.charCodeAt(0).toString(16).padStart(2, '0')}`)
        .join(''),
    )
    const claims = JSON.parse(texto) as ClaimsToken

    if (!claims.sub) {
      return null
    }

    return {
      id: claims.sub,
      email: claims.email ?? '',
      nombre: claims.name ?? claims.email ?? 'Usuario',
    }
  } catch {
    return null
  }
}
